libs/src/api/user.js

```diff
@@ -208,86 +209,74 @@ class Users extends Base {
-      this.userStateManager.setCurrentUser({ ...currentMetadata, ...metadata })
-
-      await this._updateUserOperations(metadata, currentMetadata, userId)
-
-      resp = await this.creatorNode.updateCreator(userId, metadata)
-      let updatedAudiusUserMultihash = Utils.decodeMultihash(resp.metadataMultihash)
-      if (updatedMultihashDecoded.digest !== updatedAudiusUserMultihash.digest) {
-        throw new Error(`Inconsistent multihash fields after update - ${updatedMultihashDecoded.digest} / ${updatedAudiusUserMultihash.digest}`)
-      }
-    } else {
-      throw new Error(`Cannot update creator because no current record exists for creator id ${userId}`)
+    const user = this.userStateManager.getCurrentUser()
+    if (!user) {
+      throw new Error('No current user')
     }
+    const oldMetadata = { ...user }
+
+    // Upload new metadata
+    const { metadataMultihash, metadataFileUUID } = await this.creatorNode.uploadCreatorContent(
+      newMetadata
+    )
+    // Update new metadata on chain
+    let updatedMultihashDecoded = Utils.decodeMultihash(metadataMultihash)
+    const { txReceipt } = await this.contracts.UserFactoryClient.updateMultihash(userId, updatedMultihashDecoded.digest)
+    await this._updateUserOperations(newMetadata, oldMetadata, userId)
+    if (newMetadata.creator_node_endpoint !== oldMetadata.creator_node_endpoint) {
+      await this._waitForCreatorNodeUpdate(newMetadata.user_id, newMetadata.creator_node_endpoint)
+    }
+    // Re-associate the user id with the metadata and block number
+    await this.creatorNode.updateCreator(userId, metadataFileUUID, txReceipt.blockNumber)
+
+    this.userStateManager.setCurrentUser({ ...oldMetadata, ...newMetadata })
+    return userId
   }
 
   /**
    * Upgrades a user to a creator using their metadata object.
    * This creates a record for that user on the connected creator node.
    * @param {Object} metadata
    */
-  async upgradeToCreator (metadata) {
+  async upgradeToCreator () {
     this.REQUIRES(Services.CREATOR_NODE)
-    this.IS_OBJECT(metadata)
 
-    const oldMetadata = { ...metadata }
-    this._validateUserMetadata(oldMetadata)
-
-    // for now, we only support one user per creator node / libs instance
     const user = this.userStateManager.getCurrentUser()
-    if (!user || user.is_creator) {
-      throw new Error('No current user or existing user is already a creator')
+    if (!user) {
+      throw new Error('No current user')
     }
+    // No-op if the user is already a creator.
+    // Consider them a creator iff they have is_creator=true AND a creator node endpoint
+    if (user.is_creator && user.creator_node_endpoint) return
 
-    oldMetadata.wallet = this.web3Manager.getWalletAddress()
+    const userId = user.user_id
+    const oldMetadata = { ...user }
+    const newMetadata = { ...user }
 
-    const newMetadata = { ...oldMetadata }
+    newMetadata.wallet = this.web3Manager.getWalletAddress()
```

Comment:
@raymondjacobson just a question - so these values for wallet and creator_node_endpoint are probably trigged to change somewhere in the dapp and the state has been updated in the respective services?

Reply:
There's a "setEndpoint" method that's called from the dapp. I'd like to clean up that flow a bit someday, but I think it's fair for now. Since we only allow 1 established connection to a creator node, that connection is switched/set before calling upgrade to creator